trl/trainer/online_dpo_trainer.py

```diff
@@ -0,0 +1,566 @@
+import gc
+import math
+import os
+import time
+from collections import defaultdict
+from typing import Dict, List, Optional, Tuple, Union
+
+import numpy as np
+import pandas as pd
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from accelerate import Accelerator
+from accelerate.utils import broadcast, gather_object
+from datasets import Dataset
+from torch.utils.data import DataLoader
+from transformers import (
+    DataCollatorWithPadding,
+    GenerationConfig,
+    PreTrainedTokenizer,
+    Trainer,
+    TrainerCallback,
+    TrainerControl,
+)
+from transformers.integrations import get_reporting_integration_callbacks
+from transformers.trainer import DEFAULT_CALLBACKS, DEFAULT_PROGRESS_CALLBACK
+from transformers.trainer_callback import CallbackHandler, PrinterCallback
+
+from ..models.utils import unwrap_model_for_generation
+from ..trainer.utils import (
+    OnlineTrainerState,
+    batch_generation,
+    disable_dropout_in_model,
+    exact_div,
+    first_true_indices,
+    forward,
+    get_reward,
+    prepare_deepspeed,
+    print_rich_table,
+    truncate_response,
+)
+from .online_dpo_config import OnlineDPOConfig
+
+
+INVALID_LOGPROB = 1.0
+
+
+class OnlineDPOTrainer(Trainer):
+    def __init__(
+        self,
+        config: OnlineDPOConfig,
+        tokenizer: PreTrainedTokenizer,
+        policy: nn.Module,
+        ref_policy: nn.Module,
+        reward_model: nn.Module,
+        train_dataset: Dataset,
+        data_collator: Optional[DataCollatorWithPadding] = None,
+        eval_dataset: Optional[Union[Dataset, Dict[str, Dataset]]] = None,
+        # less commonly used
+        optimizers: Tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR] = (None, None),
+        callbacks: Optional[List[TrainerCallback]] = None,
+    ) -> None:
+        self.args = config
+        args = config
+        self.tokenizer = tokenizer
+        self.policy = policy
+
+        self.policy.generation_config.eos_token_id = (
+            None  # disable `pad_token_id` and `eos_token_id` because we just want to
+        )
+        self.policy.generation_config.pad_token_id = None  # generate tokens without truncation / padding
+
+        self.ref_policy = ref_policy
+        self.reward_model = reward_model
+        self.train_dataset = train_dataset
+        self.train_dataset_len = len(train_dataset)
+        self.data_collator = data_collator
+        self.eval_dataset = eval_dataset
+        self.optimizer, self.lr_scheduler = optimizers
+
+        #########
+        # calculate various batch sizes
+        #########
+        if args.total_episodes is None:  # allow the users to define episodes in terms of epochs.
+            args.total_episodes = int(args.num_train_epochs * self.train_dataset_len)
+        accelerator = Accelerator(gradient_accumulation_steps=args.gradient_accumulation_steps)
+        self.accelerator = accelerator
+        args.world_size = accelerator.num_processes
```

Comment:
it feels slightly wrong to make these variables that aren't set by the user as `args` 
- world_size
- local_batch_size
- micro_batch_size
- batch_size
- mini_batch_size
- local_mini_batch_size
- num_updates

maybe change to `self`

Reply:
That is a good point. However, the main purpose of putting everything in `args` is so that the experiment management (e.g., wandb) could track these. After all, we'd be interested in the global effective batch size `batch_size`.

Reply:
That's a fair point. I guess an assertion that those args aren't set would be helpful but not necessary